File: typescript/ltx/src/rest.ts
/**
 * @interplanet/ltx — REST API client
 */

import { DEFAULT_API_BASE } from './constants.js';
import type { LtxPlan, LtxPlanV1 } from './types.js';

export interface SessionResponse {
  plan_id:    string;
  segments:   object[];
  total_min:  number;
  stored:     boolean;
}

export interface GetSessionResponse {
  plan_id:    string;
  plan:       LtxPlan;
  created_at: string;
  views:      number;
}

export interface FeedbackResponse {
  ok:          boolean;
  feedback_id: number;
}

/**
 * Store a session plan on the server.
 */
export async function storeSession(
  cfg: LtxPlan | LtxPlanV1,
  apiBase?: string,
): Promise<SessionResponse> {
  const url  = apiBase || DEFAULT_API_BASE;
  const resp = await fetch(`${url}?action=session`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify(cfg),
  });
  if (!resp.ok) throw new Error(`LTX API ${resp.status}: ${await resp.text()}`);
  return resp.json() as Promise<SessionResponse>;
}

/**
 * Retrieve a stored session plan by plan ID.
 */
export async function getSession(
  planId: string,
  apiBase?: string,
): Promise<GetSessionResponse> {
  const url  = apiBase || DEFAULT_API_BASE;
  const resp = await fetch(`${url}?action=session&plan_id=${encodeURIComponent(planId)}`);
  if (!resp.ok) throw new Error(`LTX API ${resp.status}: ${await resp.text()}`);
  return resp.json() as Promise<GetSessionResponse>;
}

/**
 * Download ICS content for a stored plan.
 */
export async function downloadICS(
  planId:  string,
  opts:    { start: string; duration_min: number },
  apiBase?: string,
): Promise<string> {
  const url  = apiBase || DEFAULT_API_BASE;
  const resp = await fetch(`${url}?action=ics&plan_id=${encodeURIComponent(planId)}`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify(opts),
  });
  if (!resp.ok) throw new Error(`LTX API ${resp.status}: ${await resp.text()}`);
  return resp.text();
}

/**
 * Submit session feedback.
 */
export async function submitFeedback(
  payload:  object,
  apiBase?: string,
): Promise<FeedbackResponse> {
  const url  = apiBase || DEFAULT_API_BASE;
  const resp = await fetch(`${url}?action=feedback`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify(payload),
  });
  if (!resp.ok) throw new Error(`LTX API ${resp.status}: ${await resp.text()}`);
  return resp.json() as Promise<FeedbackResponse>;
}

File: python/ltx/src/interplanet_ltx/_rest.py
"""REST API client for the LTX server — Python port of ltx-sdk.js REST methods.

Uses only stdlib urllib (no external dependencies).
"""

import json
import urllib.request
from typing import Any, Dict, Optional

from ._models import LtxPlan
from ._core import _plan_as_dict, DEFAULT_API_BASE


def _post(url: str, payload: dict) -> dict:
    data = json.dumps(payload).encode('utf-8')
    req = urllib.request.Request(
        url,
        data=data,
        headers={'Content-Type': 'application/json'},
        method='POST',
    )
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode('utf-8'))


def _get(url: str) -> dict:
    with urllib.request.urlopen(url) as resp:
        return json.loads(resp.read().decode('utf-8'))


def store_session(plan: LtxPlan, api_base: Optional[str] = None) -> Dict[str, Any]:
    """Store a session plan on the LTX server.

    Returns: {'plan_id': str, 'segments': list, 'total_min': int, 'stored': bool}
    """
    url = (api_base or DEFAULT_API_BASE) + '?action=session'
    return _post(url, _plan_as_dict(plan))


def get_session(plan_id: str, api_base: Optional[str] = None) -> Dict[str, Any]:
    """Retrieve a stored session plan by plan ID.

    Returns: {'plan_id': str, 'plan': dict, 'created_at': str, 'views': int}
    """
    base = api_base or DEFAULT_API_BASE
    from urllib.parse import quote
    url = f'{base}?action=session&plan_id={quote(plan_id)}'
    return _get(url)


def download_ics(
    plan_id: str,
    start: str,
    duration_min: int,
    api_base: Optional[str] = None,
) -> str:
    """Download ICS content for a stored plan from the server.  Returns ICS text."""
    base = api_base or DEFAULT_API_BASE
    from urllib.parse import quote
    url = f'{base}?action=ics&plan_id={quote(plan_id)}'
    data = json.dumps({'start': start, 'duration_min': duration_min}).encode('utf-8')
    req = urllib.request.Request(
        url,
        data=data,
        headers={'Content-Type': 'application/json'},
        method='POST',
    )
    with urllib.request.urlopen(req) as resp:
        return resp.read().decode('utf-8')


def submit_feedback(payload: dict, api_base: Optional[str] = None) -> Dict[str, Any]:
    """Submit session feedback.  Returns: {'ok': bool, 'feedback_id': int}"""
    url = (api_base or DEFAULT_API_BASE) + '?action=feedback'
    return _post(url, payload)
